rfcs/20190610-standartizing-composite_ops.md

```diff
@@ -0,0 +1,323 @@
+    following the moving target. Once such pattern is extracted there is a
+    pressure on op-maintainers to freeze his implementation to avoid breaking
+    the patterns used by the downstream tools. Or (c) Tring to roll out their
+    own extension such as `OpHint` for tflite.
+
+*   New ops that could push the industry forward however are stymied due to lack
+    efficient implementation, software vendors are not interested in
+    implementation until the ops become popular enough, creating a vicious
+    cycle.
+
+#### Alternative 1: Recommend each new op to be implemented as a new atomic node in graphdef.
+
+#### Pros:
+
+*   Simplified graph processing since the operation is just a single Node in the
+    graph.
+*   Clean graph def.
+*   Tooling (once supports new ops) stays forward compatible. The underlying
+    implementation is invisible.
+
+#### Cons:
+
+Introduces backward incompatible changes that could be avoided. Every time an
+operation is added, the versions of Tensorflow that would have been capable of
+processing the original graph (with core ops), will no longer be able to read
+the graph def that contains the new to process the new version, even though the
+only thing that *has* changed is the tensorflow added a faster implementation of
+that Op.
+
+For example: consider matrix exponentiation. It is implemented as a fairly
+complex tensorflow graph that uses just regular matrix multiplication and other
+standard ops. However, one can easily imagine that this implementation could be
+highly optimized if done as a single expm node, however if we replace that, it
+will break old tools, which would either need to effectively re-implement
+tensorflow original implementation.
+
+Requires custom version of tensorflow to add new ops outside of tensorflow which
+makes it out of reach for most users and introduces incompatibilities within the
+ecosystem, effectively forcing large users to be stuck at old versions of tensorflow.
+
+#### Alternative 2: Use name scopes for delineation of the ops
+
+Pros:
+
+*   Simple and intuitive graph-def
+*   Backward compatible - no new operations are added.
+
+Cons:
+
+*   very brittle for the tools to optimize such ops. If they depend on scope
+    names it can easily cause conflicts with models that are doing something
+    else or accidental renames can cause scopes to become invisible for the
+    graph processing.
+*   If tools depend on graph pattern matching, this makes it hard to change
+    implementations later on.
+
+*   Tooling is not forward compatible.
+
+## Questions and Discussion Topics
+
+1.  We should provide namespacing for the operation names early on to avoid
+    future collisions with function names. One option is to adopt java-style
+    "org.opname", basically using the developing organization as disambiguating
+    factor. Another alternative is to use semantic namespaces e.g. `linalg`.  Yet
+    third, is to use org.semantic.opname.
```

Comment:
`org.semantic.opname` for future proofing with extra disambiguation would be my vote.